otter/test/util/test_zkpartitioner.py

```diff
@@ -0,0 +1,183 @@
+        self.assertEqual(self.buckets_received, [])
+
+    def test_release(self):
+        """
+        When state is ``release``, the :obj:`SetPartitioner`'s ``release_set``
+        method is called.
+        """
+        self.kz_partitioner.release = True
+        self.partitioner.startService()
+        self.log.msg.assert_called_with('Partition changed. Repartitioning',
+                                        otter_msg_type='partition-released')
+        self.kz_partitioner.release_set.assert_called_once_with()
+        self.assertEqual(self.buckets_received, [])
+
+    def test_failed(self):
+        """When state is ``failed``, a new partitioner is created."""
+        self.kz_partitioner.allocating = True
+        self.partitioner.startService()
+
+        self.kz_partitioner.allocating = False
+        self.kz_partitioner.failed = True
+        # expect a new SetPartitioner to be created
+        new_kz_partition = object()
+        self.kz_client.SetPartitioner.return_value = new_kz_partition
+
+        self.clock.advance(10)
+
+        self.log.msg.assert_called_with('Partition failed. Starting new',
+                                        otter_msg_type='partition-failed')
+        # Called once when starting and now again when partition failed
+        self.assertEqual(self.kz_client.SetPartitioner.call_args_list,
+                         [mock.call(self.path,
+                                    set=self.buckets,
+                                    time_boundary=self.time_boundary)] * 2)
+        self.assertEqual(self.partitioner.partitioner, new_kz_partition)
+        self.assertEqual(self.buckets_received, [])
+
+    def test_invalid_state(self):
+        """
+        When none of the expected states are True, the ``finish`` method is
+        called on the partitioner and a new partitioner is created.
+        """
+        self.kz_partitioner.allocating = True
+        self.partitioner.startService()
+        self.kz_partitioner.allocating = False
+        self.kz_partitioner.state = 'bad'
+
+        # expect a new SetPartitioner to be created
+        new_kz_partition = object()
+        self.kz_client.SetPartitioner.return_value = new_kz_partition
+
+        self.clock.advance(10)
+
+        self.log.err.assert_called_with(
+            'Unknown state bad. This cannot happen. Starting new',
+            otter_msg_type='partition-invalid-state')
+        self.kz_partitioner.finish.assert_called_once_with()
+
+        # Called once when starting and now again when got bad state
+        self.assertEqual(self.kz_client.SetPartitioner.call_args_list,
+                         [mock.call(self.path,
+                                    set=self.buckets,
+                                    time_boundary=self.time_boundary)] * 2)
+        self.assertEqual(self.partitioner.partitioner, new_kz_partition)
+        self.assertEqual(self.buckets_received, [])
+
+    def test_acquired(self):
```

Comment:
Do you think one more test `test_acquired_again` that checks if buckets is received again after 10 seconds will help?